apps/rewards/app/components/Panel/NewReward/NewReward.js

```diff
@@ -76,6 +77,7 @@ const INITIAL_STATE = {
     dateEndPassed: 'End date must take place after today.',
     dateStartAfterEnd: 'Start date must take place before the end date.',
     singleDisbursement: 'While you selected a recurring dividend, based on your parameters, there will only be a single disbursement.',
+    dateBeforeAsset: 'The selected ${dateType} date occurs before the reference asset, ${tokenSymbol}, was created. Please choose another date.',
```

Comment:
While we're here, maybe we could move `messages` outside of component state, since it's never being changed. And if we restructure it a little bit, we don't need custom string-parsing logic:

```js
const messages = {
  customTokenInvalid: () => 'Token address must be of a valid ERC20 compatible clonable token.',
  meritTokenTransferable: () => 'Merit rewards must be non-transferable.',
  amountOverBalance: () => 'Amount must be below the available balance.',
  dateReferencePassed: () => 'Reference date must take place after today.',
  dateStartPassed: () => 'Start date must take place after today.',
  dateEndPassed: () => 'End date must take place after today.',
  dateStartAfterEnd: () => 'Start date must take place before the end date.',
  singleDisbursement: () => 'While you selected a recurring dividend, based on your parameters, there will only be a single disbursement.',
  dateBeforeAsset: (dateType, tokenSymbol) => `The selected ${dateType} date occurs before the reference asset, ${tokenSymbol}, was created. Please choose another date.`,
}
```

Then we can use it like this, and get rid of the whole `showMessage` function:

```js
messages[error.name](...error.params)
```

---

Backing up, I sort of wonder if this whole "using keyed message names" is worth it. Once we have i18n, maybe we would want something like this. But until then, why not push the error messages themselves directly into an `errors` array? Like, in `setSemanticErrors` (which I'd rename to `setErrors`), why not something like this:

```js
if (state.rewardType === ONE_TIME_MERIT && state.transferable) {
  errors.push('Merit rewards must be non-transferable.')
}
// etc, etc
```

Then we could get rid of the whole `messages` thing, get rid of the whole `showMessage` thing, and in our `errorBlocks` and `warningBlocks` functions, we would loop over simple arrays of strings.

Am I missing something?

Reply:
@chadoh, you're not missing anything. On the contrary, I agree with you that this logic is too complex and I appreciate your suggestions to make it better. The last part, though, I don't quite agree. I think in general strings shouldn't be part of code, and the less entangled with code, the better, so that it's easier to change them and make sure they are consistent with eachother.